### PR TITLE
widget: Add a test for receiving live events

### PR DESCRIPTION
**Pull request by jplatte**
Part of #2602.

### CODE REVIEW

**Review by daniel-abramov — APPROVED** [review]

File: crates/matrix-sdk/tests/integration/widget.rs [hunk 1]
```diff
@@ -310,9 +325,108 @@ async fn read_room_members() {
...
+    )
+    .await;
+
+    // No messages from the driver yet
+    assert_matches!(recv_message(&driver_handle).now_or_never(), None);
+
+    let mut sync_builder = SyncResponseBuilder::new();
+    // bump the internal batch counter, otherwise the response will be seen as
+    // identical to the one done in `run_test_driver`
+    sync_builder.build_json_sync_response();
```
Comment (bnjbvr) [inline crates/matrix-sdk/tests/integration/widget.rs:349]: There's also a call to `build_json_sync_response` in `run_test_driver`, shouldn't that bump the internal batch counter?

File: crates/matrix-sdk/tests/integration/widget.rs [hunk 2]
```diff
@@ -310,9 +325,108 @@ async fn read_room_members() {
...
+    negotiate_capabilities(
+        &driver_handle,
+        json!([
+            "org.matrix.msc2762.receive.event:m.room.member",
+            "org.matrix.msc2762.receive.event:m.room.message#m.text",
+            "org.matrix.msc2762.receive.state_event:m.room.name#",
+            "org.matrix.msc2762.receive.state_event:m.room.member#@example:localhost",
+        ]),
+    )
+    .await;
```
Comment (bnjbvr) [inline crates/matrix-sdk/tests/integration/widget.rs:341]: I think it would make more sense to split this, so that we can ensure that e.g. the first capability doesn't imply the third one: right now it's not clear whether an event is filtered by one or the other capability. We could iterate on those capabilities, create a new test driver/client that receives the same sync response below, and check what are the expected received messages for each given capability, instead? (or split in multiple test functions, with some helpers)

**Review by bnjbvr — CHANGES_REQUESTED** [review]

File: crates/matrix-sdk/tests/integration/widget.rs (hunk 1 above)
Comment (jplatte) [reply crates/matrix-sdk/tests/integration/widget.rs:349]: It's a different `SyncResponseBuilder` instance. I didn't want to make that function return *yet* another object.

File: crates/matrix-sdk/tests/integration/widget.rs (hunk 2 above)
Comment (jplatte) [reply crates/matrix-sdk/tests/integration/widget.rs:341]: We already have unit tests for the filters individually, as well as unit tests for the `WidgetMachine`, where we can test edge cases like that better.

File: crates/matrix-sdk/tests/integration/widget.rs (hunk 1 above)
Comment (bnjbvr) [reply crates/matrix-sdk/tests/integration/widget.rs:349]: Maybe the function could take a `&mut SyncResponseBuilder` parameter instead? Or return a struct containing all the returned fields at once, so adding a new one doesn't require modifying all the callers?

**Review by bnjbvr — APPROVED** [review]

File: crates/matrix-sdk/tests/integration/widget.rs (hunk 1 above)
Comment (jplatte) [reply crates/matrix-sdk/tests/integration/widget.rs:349]: Yeah I think this can be improved, but I'm not sure what's really the best option. The `&mut` parameter would be annoying for all the other tests.
